### PR TITLE
Expand and scramble test cases

### LINKED ISSUE #354: Expand NBS test set

### PR DESCRIPTION
**Pull request by bamader**
## Description
This PR adds a script that allow us to combinatorially expand and scramble the test cases NBS provided for us. We leave each of the original test cases unchanged, but we generate a random number of duplicates of each case to mess with. We then dropout some fields to simulate missingness, as well as apply random likelihoods of edit distances to string values in algorithm-relevant fields. All of this is controlled by parameters at the top of the script (which only needs to be run once to make the data-set, doesn't need to be run each time we run the algorithm tests). Of note, the script right now copies the match decisions of the original test case into each duplicate we generate from it, no matter how much we then scramble and mangle the duplicate case. This is the reason we have poor performance on this set, so that's expected. Before sharing this out, we should consider "re-grading" our expanded set to have better labels of when things should and shouldn't match. Also, since the script is customizable, we can also generate different test expansion cases if we want to simulate "pretty close" data (low randomness scrambling) vs "really bad data" (lots of scrambling). We'd need to grade each separately, but this could be useful for purposes of showing performance in different contexts.

## Related Issues
Closes #354 

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [ ] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [ ] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [ ] I have updated the documentation, if applicable.
- [ ] I have added or updated test cases to cover my changes, if applicable.
- [ ] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

**Comment by bamader** [comment]
Performance of the algorithm tests on this initial run

Screenshot 2025-05-08 at 2 48 25 PM

File: tests/algorithm/scripts/expand_test_data.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    data that might be relevant to the algorithm.
+    """
+    missing_fields = []
+    for field in ALGORITHM_RELEVANT_COLUMNS:
+        if test_case[field] is None or test_case[field] == "":
+            missing_fields.append(field)
+    return missing_fields
+
+
+def apply_field_scrambling(
```
Comment (ericbuckley) [inline tests/algorithm/scripts/expand_test_data.py:65]: I think this method could be really useful for some concepts I've been thinking about for load testing.  The big difference is we'll be dealing with JSON and not CSV input.  No changes necessary, but if you could think of an easy way to abstract the code so its format (json, csv, tsv, etc) agnostic, that would be really helpful.  This is also something we can iterate on in the future.

File: tests/algorithm/scripts/expand_test_data.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+test_cases = []
+with open(BASE_TEST_FILE, 'r') as fp:
+    test_cases = fp.readlines()
+
+# Now, let's do a bit of formatting and parsing:
+# Each case will be a dict mapping a formatted value to the
+# appropriate header
+headers = [h.strip() for h in test_cases[0].split(",")]
+test_cases = test_cases[1:]
+# Some field values are double-quoted strings which themselves contain commas
```
Comment (ericbuckley) [inline tests/algorithm/scripts/expand_test_data.py:180]: Just so you know, the [csv module has a reader](https://docs.python.org/3/library/csv.html#csv.reader) that does this same logic.

**Review by ericbuckley — APPROVED** [review]

File: tests/algorithm/scripts/expand_test_data.py (hunk 1 above)
Comment (bamader) [reply tests/algorithm/scripts/expand_test_data.py:65]: Yeah I suspect we'll need to iterate on this as we get into load testing, when we see what our data shapes and structures look like more precisely, but definitely agree. At its heart this is really an intelligent randomization function that keeps distributions looking reasonable and controls quality, and that's not predicated on data format. We should definitely be able to abstract that randomization to a more general case once we get into it!
